### DISCUSSION TITLE
Is there a way to reload a signed URL that still has time remaining before it expires?

### DISCUSSION THREAD

**Discussion by 2020-byte** [supabase/supabase#29011] [Questions]
Is there a way to reload a signed URL that still has time remaining before it expires using the same path? I am trying to implement a gallery where I can set images as public or private, but using createSignedUrl to fetch a new signed URL every time is inefficient in terms of performance.

**Comment by gc-ft** [comment]
You will need to cache the signed URL you retrieve and stop using it whenever it expires. There is no internal system that caches this for you, and even if there would be the round-time for that would be more or less as inefficient as creating a new one each time...

I would setup some kind of caching on the frontend for this or some obvious caching db like redis if for some reason you want to use the cache among multiple clients.

**Reply by 2020-byte** [reply]
Storing signed URLs on the frontend does not seem secure. Additionally, I want to implement my site in a serverless manner using only Next.js and Supabase. Running Redis would incur additional costs.

**Comment by GaryAustin1** [comment]
Browsers should cache signed URLs until around the point they expire.
Headers are returned to support this.
![image](https://github.com/user-attachments/assets/2de2fa53-d81c-4366-a8cc-7b24ee889906)

**Reply by 2020-byte** [reply]
How can I add an Expires header to my response? I noticed in the documentation that Cache-Control is used when uploading images. Should I be setting this header during the upload process?
<img width="668" alt="image" src="https://github.com/user-attachments/assets/1363eabc-fac9-4adc-a753-a9b9c25bad80">

**Reply by GaryAustin1** [reply]
I believe for signedURL you can't change the time.  It is the signedURL time.  It will cache it for as long as the signedURL is set to last.

**Reply by GaryAustin1** [reply]
I can confirm using the dashboard UI to generate signed URL the expire time is coming from the signedURL lifetime.  I set it to a year on a default uploaded file, and the expire went out a year.

**Reply by 2020-byte** [reply]
I understand that it is possible to set an expiration period for a signed URL, but if the signed URL cannot be cached, the URL cannot be reused every time the page is refreshed. And this is the situation I am having trouble with. (The function to get a public URL, getPublicUrl, returns the same URL each time, but the function createSignedUrl generates a new signed URL every time it is called.)

**Reply by GaryAustin1** [reply]
The expire header is used to cache in the browser.

**Reply by GaryAustin1** [reply]
![image](https://github.com/user-attachments/assets/fedbaf96-bde5-4a03-a117-6f5ac387cddd)

**Reply by gc-ft** [reply]
@GaryAustin1 I believe @2020-byte is saying that whenever you call createSignedUrl() the signed url is of course different, since the signature changes each time. If the URL changes, the browser has no idea what ETag it would be, therefor making it not send If-None-Match headers... Not sure if it would help cloudflare at least since cloudflare in the middle would see the ETag coming from supabase and might serve the cached resource from its end instead of loading the whole thing from supabase.

Either way, changing signedUrls make caching more complex and less useful.

It would require createSignedUrl() to return the same url each time until the expire happens, but this makes no sense security wise since it would need to keep track of the customer and url that was signed for that customer and return the same url for that customer only if it did not expire. Too much for the supabase backend to do internally I believe.

So yes, the meta table you suggest @2020-byte seems to be the only good way of doing what you are looking for:
1. Check meta table for signedurl for the supabase storage path, bucket and user combination.
2. If found, use that signedurl
3. If not found, createSignedUrl with 1 week expiry for example. store that url in meta table with storage path, bucket, user and the expire timestamp (probably with a timestamp a little lower than the real expire of the url, like a few minutes less or an hour less, depending a bit on step 4.) and use that new signedurl for the current load.
4. Add a cron job via pg_cron that runs every X minutes/hours/days and deletes all rows from the meta table with a expire timestamp in the past.

So setup expire timestamp to 1 hour below the actual expire of the url and run the pg_cron every 30 minutes to delete old timestamps, in order to give you a buffer of 30 minutes between expire of the cached url and removal from the meta table.

**Reply by GaryAustin1** [reply]
Yes if you issue a new signedURL then it will not be cached initially and there is no way to extend the same signedURL.
